### PR TITLE
Add URL validator

### LINKED ISSUE #46: Spaces in liquid variables in relative link sources result in text interpretation
Liquid has a syntax that wraps variables in double moustache braces that
may or may not have spaces within the moustaches. For example, to get the
link of the page root, you would use `{{ page.root }}` to make it more
readable. However, this violates the expectation of the commonmark parser
and makes it think “oh, this is just ordinary text”. For example, here are
two relative links, one that works, and one that is treated as text

``` r
x <- "[a] and [b]\n\n[a]: https://example.com\n[b]: {{ page.root }}/example"
cat(commonmark::markdown_xml(x))
#> <?xml version="1.0" encoding="UTF-8"?>
#> <!DOCTYPE document SYSTEM "CommonMark.dtd">
#> <document xmlns="http://commonmark.org/xml/1.0">
#>   <paragraph>
#>     <link destination="https://example.com" title="">
#>       <text xml:space="preserve">a</text>
#>     </link>
#>     <text xml:space="preserve"> and [b]</text>
#>   </paragraph>
#>   <paragraph>
#>     <text xml:space="preserve">[b]: {{ page.root }}/example</text>
#>   </paragraph>
#> </document>
```

This becomes a problem if the liquid syntax appears before a valid link:

``` r
x <- "[a] and [b]\n\n[a]: {{ page.root }}/example\n[b]: https://example.com"
cat(commonmark::markdown_xml(x))
#> <?xml version="1.0" encoding="UTF-8"?>
#> <!DOCTYPE document SYSTEM "CommonMark.dtd">
#> <document xmlns="http://commonmark.org/xml/1.0">
#>   <paragraph>
#>     <text xml:space="preserve">[a] and [b]</text>
#>   </paragraph>
#>   <paragraph>
#>     <text xml:space="preserve">[a]: {{ page.root }}/example</text>
#>     <softbreak />
#>     <text xml:space="preserve">[b]: https://example.com</text>
#>   </paragraph>
#> </document>
```

Now it is all text and nothing is a link. One solution is to just remove
the spaces:

``` r
y <- gsub("[{] ", "{", x)
y <- gsub(" [}]", "}", y)
cat(commonmark::markdown_xml(y))
#> <?xml version="1.0" encoding="UTF-8"?>
#> <!DOCTYPE document SYSTEM "CommonMark.dtd">
#> <document xmlns="http://commonmark.org/xml/1.0">
#>   <paragraph>
#>     <link destination="{{page.root}}/example" title="">
#>       <text xml:space="preserve">a</text>
#>     </link>
#>     <text xml:space="preserve"> and </text>
#>     <link destination="https://example.com" title="">
#>       <text xml:space="preserve">b</text>
#>     </link>
#>   </paragraph>
#> </document>
```

<sup>Created on 2021-07-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### PR DESCRIPTION
**Pull request by zkamvar**
This PR demonstrates first drafts of link and image validation schemes in the Episode class. The changes are listed below:

This will fix #46 and address #38. This does NOT yet address #39, as it will take more engineering and careful thought about calling out to links, but we have the structure to potentially accomplish that now. 

## NEW FEATURES

-   The Episode class now has the `$links` and `$images` active bindings that
    extracts the links and images (markdown and HTML) from the document.
-   `make_link_table()` creates a table of links parsed via `xml2::url_parse()`
    with additional information about caption and alternative text (for images).
-   The Episode class now has the `$validate_links()` method, which will validate
    links and images for common errors.
-   `Episode$use_sandpaper()` now converts images to use alt text over captions.
    Images that had `alt` are converted to `{alt='alt'}` because
    pandoc uses everything in square brackets to be caption text. NOTE: this now
    makes a copy of the XML document.

## BUG FIXES

-   `Episode$new()` gains the argument `fix_liquid`, which fixes liquid variables
    in relative links before they are passed to {tinkr}
    (<https://github.com/carpentries/pegboard/issues/46>)
-   Post-processed links via `fix_links()` retains their `sourcepos` attribute
    (fixed in e5508cc9c9a3821381293bdac12647edfbc0608e).
-   `Episode$lesson` no longer assumes that the episode is inside a sub-folder
    (fixed in 63432ef83ecc41a6aab53fe768e8eaec107278d5).

## DEMO

This demonstrates the new features for the reporting:

``` r
library("pegboard")
cats <- system.file("tests/testthat/examples/image-test.md", package = "pegboard")
link <- system.file("tests/testthat/examples/link-test.md", package = "pegboard")
cats <- Episode$new(cats)
link <- Episode$new(link)
cli::rule("Validators for images")
#> ── Validators for images ───────────────────────────────────────────────────────
cats$show()
#> ---
#> title: Markdown Images
#> ---
#> 
#> ## Image with alt text
#> 
#> ![image 1 caption](https://placekitten.com/g/101/101){alt="a small kitten"}
#> 
#> ## Image with only caption
#> 
#> ![image 2 caption](https://placekitten.com/g/102/102)
#> 
#> ## Image with no caption, but has alt text
#> 
#> ![](https://placekitten.com/g/103/103){alt="a small kitten"}
#> 
#> ## Image with other attributes, including classes
#> 
#> ![](https://placekitten.com/g/104/104){.image-with-shadow alt="a small kitten"}
#> 
#> ## Image with wrapping alt text
#> 
#> ![](https://placekitten.com/g/105/105){.image-with-shadow alt="a small kitten
#> generated by placekitten.com/"}
#> 
#> ## Image in HTML block
#> 
#> <img 
#> src="https://placekitten.com/g/106/106" 
#> alt="a small kitten"
#> />
#> 
#> ## Image in HTML inline
#> 
#> This is an image inline <img src="https://placekitten.com/g/107/107" alt="a small kitten"/>, but really shouldn't be allowed.
#> 
#> ## Multiple images in an HTML block
#> 
#> <img src="https://placekitten.com/g/108/108" alt="a small kitten"/>
#> <img src="https://placekitten.com/g/109/109"/>
cats$validate_links()
#> ! Images need alt-text:
#> https://placekitten.com/g/102/102 (image-test.md:12)
#> https://placekitten.com/g/109/109 (image-test.md:41)
#> enforce_https internal_okay all_reachable  img_alt_text   descriptive 
#>          TRUE          TRUE          TRUE         FALSE          TRUE
cli::rule("Validators for links")
#> ── Validators for links ────────────────────────────────────────────────────────
link$show()
#> ---
#> title: Link Tests
#> ---
#> 
#> ## Internal links {#internal}
#> 
#> ### :crystal\_ball: Heading with Emoji
#> 
#> ### Heading with long name that has a label {#label-1}
#> 
#> ### Another heading with a long name that has a label {#label-2}
#> 
#> ### Heading with a class, but no label {.challenge}
#> 
#> ### Heading with class and a label {.solution #label-3}
#> 
#> This is a [link to the Heading with Emoji](#heading-with-emoji) and a [link to
#> label 1][rel-label-1] and a [link](#label-2) and [this link](#label-2).
#> 
#> This link goes to [the heading with class and a label](#label-3).
#> 
#> This link is [absolutely incorrect](#bad-fragment)
#> 
#> This [relative link goes to label 1][rel-label-1]
#> 
#> ## Cross-Lesson links {#cross-lesson}
#> 
#> This [link will go to the image test][rel-image], but [this link is
#> wrong](incorrect-link.html)
#> 
#> [This link also goes to image test](image-test)
#> [This link also goes to image test, but with a slash](image-test/), though this
#> may not work for us because it implies that there is an `index.html` hiding in
#> there.
#> 
#> This link [should be a relative link](rel-image).
#> This link [is a relative link that works][rel-image]
#> 
#> ## HTTP links
#> 
#> This [link uses http, which is no bueno](http://example.com)
#> 
#> [rel-label-1]: #label-1
#> [rel-image]: image-test.html
link$validate_links()
#> ! Link text should be more descriptive than 'link':
#> 'link' (link-test.md:18)
#> 'this link' (link-test.md:18)
#> ! Links must use HTTPS, not HTTP:
#> http://example.com (link-test.md:41)
#> ! The following anchors do not exist in the file:
#> #bad-fragment (link-test.md:22)
#> ! Relative links that are incorrectly formatted:
#> [should be a relative link](rel-image) -> [should be a relative link][rel-image] (link-test.md:36)
#> ! These files do not exist in the lesson:
#> incorrect-link.html (link-test.md:29)
#> enforce_https internal_okay all_reachable  img_alt_text   descriptive 
#>         FALSE         FALSE          TRUE          TRUE         FALSE
```

<sup>Created on 2021-07-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### CODE REVIEW

**Comment by zkamvar** [comment]
Notes for first pass:

 - [ ] validate presence of alt text
 - [ ] validate https over http
 - [ ] validate internal links work
 - [ ] check phantom links

**Comment by zkamvar** [comment]
Testing this on a real live lesson, we can see the context for me opening https://github.com/data-lessons/instructor-training/pull/95 and https://github.com/data-lessons/instructor-training/pull/96:

``` r
library("pegboard")
l <- get_lesson("data-lessons/instructor-training", fix_liquid = TRUE)
#> Loading required namespace: gert
res <- purrr::map(l$episodes, ~.x$unblock()$use_sandpaper()$validate_links())
#> ! These files do not exist in the lesson:
#> 20-live/ (05-memory.md:275)
#> ! Relative links that are incorrectly formatted:
#> [Core Values](core-values) -> [Core Values][core-values] (09-eia.md:46)
res[!purrr::map_lgl(res, all)]
#> $`05-memory.md`
#> enforce_https internal_okay all_reachable  img_alt_text   descriptive 
#>          TRUE         FALSE          TRUE          TRUE          TRUE 
#> 
#> $`09-eia.md`
#> enforce_https internal_okay all_reachable  img_alt_text   descriptive 
#>          TRUE         FALSE          TRUE          TRUE          TRUE
```

<sup>Created on 2021-07-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

File: R/validate_links.R [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  if (any(in_page)) {
+    our_headings <- clean_headings(headings)
+    no_match <- !lt$fragment %in% our_headings
+    res <- res & !any(in_page & no_match)
+  } else {
+    no_match <- FALSE
+  }
+  if (any(cross_page)) {
+    neighbors <- get_lesson_files(yrn$lesson)
+    no_file <- !lt$path %in% glue::glue("{neighbors}")
```
Comment (fmichonneau) [inline R/validate_links.R:95]: The comparison of the paths here means that image path written as `./img/file.png` for instance, are marked as missing.

Comment (zkamvar) [reply R/validate_links.R]: I've addressed this in 08d7746

File: tests/testthat/_snaps/validation.md [hunk 2]
```diff
@@ -127,7 +165,7 @@
     Message <cliMessage>
       ────────────────────────────────────────────────────────────────────────────────
 
-# duplciate reporting works [plain]
+# duplciate headings reporting works [plain]
```
Comment (fmichonneau) [inline tests/testthat/_snaps/validation.md:168]:
```suggestion
# duplicate headings reporting works [plain]
```

File: tests/testthat/_snaps/validation.md [hunk 3]
```diff
@@ -149,7 +187,7 @@
     Message <cliMessage>
       --------------------------------------------------------------------------------
 
-# duplciate reporting works [ansi]
+# duplciate headings reporting works [ansi]
```
Comment (fmichonneau) [inline tests/testthat/_snaps/validation.md:190]:
```suggestion
# duplicate headings reporting works [ansi]
```

File: tests/testthat/_snaps/validation.md [hunk 4]
```diff
@@ -171,7 +209,7 @@
     Message <cliMessage>
       --------------------------------------------------------------------------------
 
-# duplciate reporting works [unicode]
+# duplciate headings reporting works [unicode]
```
Comment (fmichonneau) [inline tests/testthat/_snaps/validation.md:212]:
```suggestion
# duplicate headings reporting works [unicode]
```

File: tests/testthat/_snaps/validation.md [hunk 5]
```diff
@@ -193,7 +231,7 @@
     Message <cliMessage>
       ────────────────────────────────────────────────────────────────────────────────
 
-# duplciate reporting works [fancy]
+# duplciate headings reporting works [fancy]
```
Comment (fmichonneau) [inline tests/testthat/_snaps/validation.md:234]:
```suggestion
# duplicate headings reporting works [fancy]
```

File: tests/testthat/test-validation.R [hunk 6]
```diff
@@ -1,26 +1,49 @@
...
+})
+
 if (requireNamespace("cli", quietly = TRUE)) {
-  cli::test_that_cli("reporters will work", {
+  cli::test_that_cli("headings reporters will work", {
     expect_snapshot(expect_false(all(vh$validate_headings())))
   })
 
-  cli::test_that_cli("duplciate reporting works", {
+  cli::test_that_cli("duplciate headings reporting works", {
```
Comment (fmichonneau) [inline tests/testthat/test-validation.R:40]:
```suggestion
  cli::test_that_cli("duplicate headings reporting works", {
```

**Review by fmichonneau — APPROVED** [review]
this looks very good! Thanks for working on it. 

I noticed that R/fix_liquid.R is an empty file, I don't know if it's on purpose.
There is also a small typo in the tests.

**Comment by zkamvar** [comment]
> I noticed that R/fix_liquid.R is an empty file, I don't know if it's on purpose. 

Definitely NOT on purpose! I will fix it :)

**Pull request merged**